# Проект: Исследование надежности заемщиков банка

**Оглавление**<a id='toc0_'></a>    
- 1. [Загрузка и изучение данных](#toc1_)    
  - 1.1. [Вывод по результатам раздела](#toc1_1_)    
- 2. [Предобработка данных](#toc2_)    
  - 2.1. [Удаление пропусков](#toc2_1_)    
  - 2.2. [Обработка аномальных значений](#toc2_2_)    
  - 2.3. [Удаление пропусков (продолжение)](#toc2_3_)    
  - 2.4. [Изменение типов данных](#toc2_4_)    
  - 2.5. [Обработка дубликатов](#toc2_5_)    
  - 2.6. [Категоризация данных](#toc2_6_)    
- 3. [Исследование данных и ответы на вопросы](#toc3_)    
  - 3.1. [Есть ли зависимость между количеством детей и возвратом кредита в срок?](#toc3_1_)    
  - 3.2. [Есть ли зависимость между семейным положением и возвратом кредита в срок?](#toc3_2_)    
  - 3.3. [Есть ли зависимость между уровнем дохода и возвратом кредита в срок?](#toc3_3_)    
  - 3.4. [Как разные цели кредита влияют на его возврат в срок?](#toc3_4_)    
  - 3.5. [Приведите возможные причины появления пропусков в исходных данных.](#toc3_5_)    
  - 3.6. [Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.](#toc3_6_)    
- 4. [Общий вывод](#toc4_)    

<!-- vscode-jupyter-toc-config
	numbering=true
	anchor=true
	flat=false
	minLevel=2
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

**Задание:**

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели *кредитного скоринга* — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

План работы:
1. Загрузка и изучение данных
2. Предобработка данных
3. Исследование данных и ответы на вопросы
4. Общий вывод


> Часть пунктов проекта оформлены в виде выполненных заданий на платформа Яндекс Практикум, включенных в общий проект.

## 1. <a id='toc1_'></a>[Загрузка и изучение данных](#toc0_)

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd
from IPython.display import display

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### 1.1. <a id='toc1_1_'></a>[Вывод по результатам раздела](#toc0_)

Выполнена загрузка и первичное исследование данных. Имеющихся данных для исследования достаточно (более 20 тыс. строк), но необходимо выполнить их первичную обработку по удалению и корректировке пропусков, дубликатов и некорректных данных.

## 2. <a id='toc2_'></a>[Предобработка данных](#toc0_)

### 2.1. <a id='toc2_1_'></a>[Удаление пропусков](#toc0_)

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### 2.2. <a id='toc2_2_'></a>[Обработка аномальных значений](#toc0_)

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### 2.3. <a id='toc2_3_'></a>[Удаление пропусков (продолжение)](#toc0_)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### 2.4. <a id='toc2_4_'></a>[Изменение типов данных](#toc0_)

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### 2.5. <a id='toc2_5_'></a>[Обработка дубликатов](#toc0_)

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### 2.6. <a id='toc2_6_'></a>[Категоризация данных](#toc0_)

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## 3. <a id='toc3_'></a>[Исследование данных и ответы на вопросы](#toc0_)

### 3.1. <a id='toc3_1_'></a>[Есть ли зависимость между количеством детей и возвратом кредита в срок?](#toc0_)

Для определения зависимости между количеством детей и возвратом кредита в срок, мы можем опираться на имеющийся в выборке показатель имевших проблемы с задолженностью (поле "debt" — имел ли задолженность по возврату кредитов).
Определим общий процент заемщиков, имевших задолженностью по возврату кредита, для всей базы и сравним его с долей имевших задолженность, сгруппированных по количеству имеющихся у них детей.

In [23]:
print('1) Раcпределение заемщиков по группам в зависимости от количества имеющихся у них детей.')
print(data['children'].value_counts(), '\n')
print('2) Количество имевших задолженность сгруппированных по количеству имеющихся у них детей.\n', data.groupby('children')['debt'].sum(), '\n')
debt_tot = data['debt'].sum() / data['debt'].count()
print(f'3) Общий процент заемщиков, имевших задолженность по возврату кредита, для всей базы составляет {debt_tot:.2%}.\n')
print('4) Доля имевших задолженность по кредиту для групп заемщиков в зависимости от количества детей:')
zav_children = (data.groupby('children')['debt'].sum() / data['children'].value_counts())
for i in range(len(zav_children)):
    print(f'Количество детей: {i}. Количество, имевших задолженность: {zav_children[i]:.2%}.')

1) Раcпределение заемщиков по группам в зависимости от количества имеющихся у них детей.
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64 

2) Количество имевших задолженность сгруппированных по количеству имеющихся у них детей.
 children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64 

3) Общий процент заемщиков, имевших задолженность по возврату кредита, для всей базы составляет 8.12%.

4) Доля имевших задолженность по кредиту для групп заемщиков в зависимости от количества детей:
Количество детей: 0. Количество, имевших задолженность: 7.54%.
Количество детей: 1. Количество, имевших задолженность: 9.23%.
Количество детей: 2. Количество, имевших задолженность: 9.45%.
Количество детей: 3. Количество, имевших задолженность: 8.18%.
Количество детей: 4. Количество, имевших задолженность: 9.76%.
Количество детей: 5. Количество, имевших задолженность: 0.00%.


**Вывод:** 
1. Общий процент заемщиков, имевших проблемы с возвратом задолженности, составляет 8,12% от базы. При этом процент имевших задолженность в группах с количеством детей от 1 до 4 находится в диапазоне от 8,18% до 9,76%. Превышение общего процента незначительное, можно предполагать, что существенной зависимости между возвратом кредита в срок и количеством детей у заемщика нет.
2. В группе заемщиков, имеющих 5 детей, имевших задолженность нет. Но в этой группе всего 9 человек (сотые доли процента от общей базы), что говорит о нерепрезентативности выборки и невозможности сделать однозначный вывод.
3. В группе заемщиков, не имеющих детей, доля, имевших задолженность составляет 7,54%. Это незначительно меньше, общего процента имевших задолженность, поэтому также можно предположить, что существенной зависимости между возвратом кредита в срок и наличием у заемщика детей нет.

Теперь соберем все данные в общую таблицу, проанализируем и сделаем выводы.

In [24]:
data_pivot = data.pivot_table(values='debt', index='children', aggfunc=['count','sum', 'mean'], margins=True)
data_pivot['mean'] = data_pivot['mean'].round(4) * 100
data_pivot['Отклонение от общего процента должников, %'] = data_pivot['mean'] - data_pivot.loc['All', 'mean']
data_pivot.rename(columns={'count': 'Количество заемщиков', 'sum': 'Количество должников', 'mean': 'Средний процент должников, %', 'debt': ''}, inplace=True)
data_pivot.index.names = ['Количество детей']
display(data_pivot)

Количество заемщиков Количество должников  \
                                                             
Количество детей                                             
0                               14091                 1063   
1                                4808                  444   
2                                2052                  194   
3                                 330                   27   
4                                  41                    4   
5                                   9                    0   
All                             21331                 1732   

                 Средний процент должников, %  \
                                                
Количество детей                                
0                                        7.54   
1                                        9.23   
2                                        9.45   
3                                        8.18   
4                                        9.76   
5                                        0.00   
All                                      8.12   

                 Отклонение от общего процента должников, %  
                                                             
Количество детей                                             
0                                                     -0.58  
1                                                      1.11  
2                                                      1.33  
3                                                      0.06  
4                                                      1.64  
5                                                     -8.12  
All                                                    0.00

**Вывод:** 

Заемщики, имеющие трёх и более детей составляют менее 2% от базы. Их влияние незначительно. Основная доля заемщиков или не имеет детей, или имеет одного или двух детей.<p>
Средний процент должников среди заемщиков, не имеющих детей, ниже общего процента должников на 0,58%.<p>
Средний процент должников среди заемщиков, имеющих 1 и 2 детей, выше общего процента должников на 1,11% и 1,33% соответственно.<p>
Можно сделать вывод, что заемщики, не имеющие детей, лучше возвращают кредиты.<p>

### 3.2. <a id='toc3_2_'></a>[Есть ли зависимость между семейным положением и возвратом кредита в срок?](#toc0_)

Распределим заемщиков на группы по семейному положению, определим процент имевших задолженность в каждой группе и сравним с общим процентом имевших задолженность (8,12%).

In [25]:
print('1) Группы заемщиков по семейному положению.')
print(data['family_status'].value_counts(), '\n')
print('2) Доля заемщиков, имевших задолженность по кредиту, в группах по семейному положению.')
print(data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['debt'].count())

1) Группы заемщиков по семейному положению.
женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64 

2) Доля заемщиков, имевших задолженность по кредиту, в группах по семейному положению.
family_status
Не женат / не замужем    0.097639
в разводе                0.070648
вдовец / вдова           0.066246
гражданский брак         0.093130
женат / замужем          0.075606
Name: debt, dtype: float64


**Вывод:** 
В группах заемщиков со статусом "вдовец / вдова", "в разводе" и "женат / замужем" доля имевших задолженность ниже общего процента и составляет 6,62%, 7,06% и 7,56% соответственно.
В группах заемщиков со статусом "Не женат / не замужем" и "гражданский брак" доля имевших задолженность выше общего процента и составляет 9,79% и 9,31% соответственно.
В целом эти отклонения незначительны, поэтому можно предположить, что зависимость возврата кредита от семейного положения не существенна.

Соберем все данные в одну таблицу, проанализируем и сделаем выводы.

In [26]:
data_pivot = data.pivot_table(values='debt', index='family_status', aggfunc=['count','sum', 'mean'], margins=True)
data_pivot['mean'] = data_pivot['mean'].round(4) * 100
data_pivot['Отклонение от общего процента должников, %'] = data_pivot['mean'] - data_pivot.loc['All', 'mean']
data_pivot.rename(columns={'count': 'Количество заемщиков', 'sum': 'Количество должников', 'mean': 'Средний процент должников, %', 'debt': ''}, inplace=True)
data_pivot.index.names = ['Семейное положение']
display(data_pivot)

Количество заемщиков Количество должников  \
                                                                  
Семейное положение                                                
Не женат / не замужем                 2796                  273   
в разводе                             1189                   84   
вдовец / вдова                         951                   63   
гражданский брак                      4134                  385   
женат / замужем                      12261                  927   
All                                  21331                 1732   

                      Средний процент должников, %  \
                                                     
Семейное положение                                   
Не женат / не замужем                         9.76   
в разводе                                     7.06   
вдовец / вдова                                6.62   
гражданский брак                              9.31   
женат / замужем                               7.56   
All                                           8.12   

                      Отклонение от общего процента должников, %  
                                                                  
Семейное положение                                                
Не женат / не замужем                                       1.64  
в разводе                                                  -1.06  
вдовец / вдова                                             -1.50  
гражданский брак                                            1.19  
женат / замужем                                            -0.56  
All                                                         0.00

**Вывод:** 

Заемщики категорий "женат / замужем", "в разводе", "вдовец / вдова" возвращают кредиты лучше, на 0,56%, 1,06% и 1,5% соответственно. При этом категория "женат / замужем" является самой многочисленной, значительно превышающей все остальные. Категории "в разводе" и "вдовец / вдова" являются самыми малочисленными.<p>
Заемщики категорий "гражданский брак" и "Не женат / не замужем" возвращают кредиты хуже, на 1,19% и 1,64% соответственно. По количеству заемщиков данные категории примерно равновеликие.<p>


### 3.3. <a id='toc3_3_'></a>[Есть ли зависимость между уровнем дохода и возвратом кредита в срок?](#toc0_)

Все клиенты разбиты на категории в зависимости от уровня их дохода:

`'A'` — 1000001 и выше;

`'B'` — 200001–1000000;

`'C'` — 50001–200000;

`'D'` — 30001–50000;

`'E'` — 0–30000.

Определим долю имевших задолженность в каждой категории и сравним с общим процентом имевших задолженность (8,12%).

In [27]:
print('1) Количество клиентов по категориям дохода.')
print(data.groupby('total_income_category')['debt'].count(), '\n')
print('2) Доля клиентов, имевших задолженность по кредиту, в разбивке по категориям дохода.')
categorys = pd.Series(index=['A', 'B', 'C', 'D', 'E'], data=['1000001 и выше', '200001–1000000', '50001–200000', '30001–50000', '0–30000'])
procent_dolzhnikov = data.groupby('total_income_category')['debt'].sum() / data.groupby('total_income_category')['debt'].count()
for i in range(len(categorys)):
    print('В категории \"{0}\" доля имеющих задолженность составляет {1:.2%}.'. format(categorys[i], procent_dolzhnikov[i]))

1) Количество клиентов по категориям дохода.
total_income_category
A       25
B     5014
C    15921
D      349
E       22
Name: debt, dtype: int64 

2) Доля клиентов, имевших задолженность по кредиту, в разбивке по категориям дохода.
В категории "1000001 и выше" доля имеющих задолженность составляет 8.00%.
В категории "200001–1000000" доля имеющих задолженность составляет 7.06%.
В категории "50001–200000" доля имеющих задолженность составляет 8.50%.
В категории "30001–50000" доля имеющих задолженность составляет 6.02%.
В категории "0–30000" доля имеющих задолженность составляет 9.09%.


**Вывод:** 

В категориях дохода "30001–50000", "200001–1000000", "1000001 и выше" доля имевших задолженность ниже среднего, в категориях "0–30000", "50001–200000" выше среднего. Все отклонения незначительны.

Соберем все данные в одну таблицу, проанализируем и сделаем выводы.

In [28]:
data_pivot = data.pivot_table(values='debt', index='total_income_category', aggfunc=['count','sum', 'mean'], margins=True)
data_pivot['mean'] = data_pivot['mean'].round(4) * 100
data_pivot['Отклонение от общего процента должников, %'] = data_pivot['mean'] - data_pivot.loc['All', 'mean']
data_pivot.insert(0, 'Уровень дохода, руб.', ['1000001 и выше', '200001–1000000', '50001–200000', '30001–50000', '0–30000', ''])
data_pivot.rename(columns={'count': 'Количество заемщиков', 'sum': 'Количество должников', 'mean': 'Средний процент должников, %', 'debt': ''}, inplace=True)
data_pivot.index.names = ['Категория']
display(data_pivot)

Уровень дохода, руб. Количество заемщиков Количество должников  \
                                                                           
Категория                                                                  
A               1000001 и выше                   25                    2   
B               200001–1000000                 5014                  354   
C                 50001–200000                15921                 1353   
D                  30001–50000                  349                   21   
E                      0–30000                   22                    2   
All                                           21331                 1732   

          Средний процент должников, %  \
                                         
Категория                                
A                                 8.00   
B                                 7.06   
C                                 8.50   
D                                 6.02   
E                                 9.09   
All                               8.12   

          Отклонение от общего процента должников, %  
                                                      
Категория                                             
A                                              -0.12  
B                                              -1.06  
C                                               0.38  
D                                              -2.10  
E                                               0.97  
All                                             0.00

**Вывод:** 

Основными категориями по количеству заемщиков являются:<p>
- категория "C" с уровнем дохода 50001–200000 руб.;<p>
- категория "B" с уровнем дохода 200001–1000000 руб.<p>

В категории "C" возврат кредитов осуществляется хуже среднего на 0,38%.<p>
В категории "B" возврат кредитов осуществляется лучше среднего на 1,06%.<p>

### 3.4. <a id='toc3_4_'></a>[Как разные цели кредита влияют на его возврат в срок?](#toc0_)

Клиенты в зависимости от цели кредита разбиты на следующие категории:
- операции с автомобилем,
- операции с недвижимостью,
- проведение свадьбы,
- получение образования.

Определим долю имеющих задолженность в каждой категории и сравним со средним значением по базе (8,12%).

In [29]:
# Собираем все данные в датафрейм
celi = pd.DataFrame(data=data.groupby('purpose_category')['debt'].count())
celi = celi.rename(columns={'debt': 'Клиенты'})
celi['Задолженность, %'] = (data.groupby('purpose_category')['debt'].sum() / data.groupby('purpose_category')['debt'].count())*100
celi.index.names = ['Цель кредита:']
celi['Задолженность, %'] = celi['Задолженность, %'].round(2)
celi['Отклонение, %'] = celi['Задолженность, %'] - 8.12
celi = celi.sort_values(by='Отклонение, %')
print(celi)

                          Клиенты  Задолженность, %  Отклонение, %
Цель кредита:                                                     
операции с недвижимостью    10751              7.26          -0.86
проведение свадьбы           2313              7.91          -0.21
получение образования        3988              9.25           1.13
операции с автомобилем       4279              9.35           1.23


**Вывод:** 

Клиенты, взявшие кредит на цели, связанные с недвижимостью и свадьбами, имеют меньший процент задолженностей. Клиенты, взявшие кредит на цели, связанные с образованием и автомобилем, имеют больший процент задолженностей.

### 3.5. <a id='toc3_5_'></a>[Приведите возможные причины появления пропусков в исходных данных.](#toc0_)

*Ответ:* 
1) Человеческий фактор (ошибки ручного ввода).

2) Технологические сбои при передаче, копировании, форматировании данных.

### 3.6. <a id='toc3_6_'></a>[Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.](#toc0_)

*Ответ:* 

Этот метод исключает некорректное влияние малочисленных, но сильно удаленных от среднего значений.

## 4. <a id='toc4_'></a>[Общий вывод](#toc0_)

В проекте использованы исходные данные банка со статистикой платежеспособности клиентов (21525 записей). Успешно проведены все этапы предобработки данных статистики платежеспособности клиентов и на полученном выверенном массиве проверено несколько гипотез, необходимых для построения модели кредитного скроллинга банка.

На этапе предобработки данных проведены следующие мероприятия:

1. Заполнение пропусков данных по уровню дохода и трудовому стажу для 2174 записей медианными значениями соответствующего типа занятости заемщика.
    
2. Обработка аномальных значений в данных:
    
    - исправление отрицательных значений трудового стажа;
        
    - удаление записей с отрицательным и некорректно большим количеством детей (123 записи);
        
    - аномально большой стаж в категориях "безработные" и "пенсионеры" не корректировался, т.к. эти данные не участвуют и не влияют на исследования.
        
3. Исключение дубликатов (71 запись).
    
4. Категоризация данных по уровню дохода и цели кредитования заемщиков.
    
На этапе исследования проверены следующие гипотезы:

1. Зависимость возврата кредита в срок от количества детей.
    
    Выводы:
        
    - Средний процент должников среди заемщиков, не имеющих детей, ниже общего процента должников на 0,58%.
        
    - Средний процент должников среди заемщиков, имеющих 1 и 2 детей, выше общего процента должников на 1,11% и 1,33% соответственно.
        
    - Влияние остальных категорий незначительно.
        
2. Зависимость возврата кредита в срок от семейного положения.
    
    Выводы:
        
    - Заемщики категорий "женат / замужем", "в разводе", "вдовец / вдова" возвращают кредиты лучше, на 0,56%, 1,06% и 1,5% соответственно. При этом категория "женат / замужем" является самой многочисленной, значительно превышающей все остальные. Категории "в разводе" и "вдовец / вдова" являются самыми малочисленными.
        
    - Заемщики категорий "гражданский брак" и "Не женат / не замужем" возвращают кредиты хуже, на 1,19% и 1,64% соответственно. По количеству заемщиков данные категории примерно равновеликие.
        
3. Зависимость возврата кредита в срок от уровня дохода.
    
    Выводы:
        
    - В категории "C" (уровень дохода 50001–200000 руб.) возврат кредитов осуществляется хуже среднего на 0,38%.
        
    - В категории "B" (уровень дохода 200001–1000000 руб.) возврат кредитов осуществляется лучше среднего на 1,06%.
        
    - Влияние остальных категорий незначительно.
        
4. Зависимость возврата кредита в срок от цели кредитования.
    
    Выводы:
        
    - Клиенты, взявшие кредит на цели, связанные с недвижимостью и свадьбой, возвращают кредиты лучше на 0,86% и 0,21% соответственно.
        
    - Клиенты, взявшие кредит на цели, связанные с образованием и автомобилем, возвращают кредиты хуже на 1,13% и 1,23% соответственно.
        
Результаты исследования планируется учесть при построении модели кредитного скоринга.